engine/configurator/conf_cyaml.c

```diff
@@ -0,0 +1,796 @@
+        case CYAML_OK:
+            return 0;
+
+        /* Memory allocation failed. */
+        case CYAML_ERR_OOM:
+            ERROR("There is a CYAML error %s(%d)", cyaml_strerror(err), err);
+            return TE_ENOMEM;
+
+        default:
+        {
+            ERROR("There is a CYAML error %s(%d)", cyaml_strerror(err), err);
+            return TE_EINVAL;
+        }
+    }
+}
+
+void
+cyaml_free_hist_seq(history_seq hist_seq);
+
+void
+cyaml_free_obj(object_type obj)
+{
+    unsigned int i;
+
+    free(obj.d);
+    free(obj.oid);
+    free(obj.def_val);
+    for (i = 0; i < obj.depends_count; i++)
+        free(obj.depends[i].oid);
+    free(obj.depends);
+}
+
+void
+cyaml_free_inst(instance_type inst)
+{
+    free(inst.if_cond);
+    free(inst.oid);
+    free(inst.value);
+}
+
+
+void
+cyaml_free_b_entry(backup_entry entry)
+{
+
+    if (entry.object != NULL)
+    {
+        cyaml_free_obj(*entry.object);
+        free(entry.object);
+    }
+
+    if (entry.instance != NULL)
+    {
+        cyaml_free_inst(*entry.instance);
+        free(entry.instance);
+    }
+}
+
+void
+cyaml_free_b_seq(backup_seq b_seq)
+{
+    unsigned int i;
+
+    for (i = 0; i < b_seq.entries_count; i++)
+        cyaml_free_b_entry(b_seq.entries[i]);
+    free(b_seq.entries);
+    b_seq.entries = NULL;
+    b_seq.entries_count = 0;
+}
+
+void
+cyaml_free_cond_entry(cond_entry cond)
+{
+    free(cond.if_cond);
+    if (cond.then_cond != NULL)
+    {
+        cyaml_free_hist_seq(*cond.then_cond);
+        free(cond.then_cond);
+    }
+    if (cond.else_cond != NULL)
+    {
+        cyaml_free_hist_seq(*cond.else_cond);
+        free(cond.else_cond);
+    }
+}
+
+void
+cyaml_free_hist_entry(history_entry entry)
+{
+    unsigned int i;
+
+    free(entry.comment);
+
+    if (entry.cond != NULL)
+    {
+        cyaml_free_cond_entry(*entry.cond);
+        free(entry.cond);
+    }
+
+    free(entry.reboot_ta);
+
+#define FREE_SEQUENCE(cmd_, free_func_) \
+    do {                                              \
+        for (i = 0; i < entry.cmd_ ## _count; i++)    \
+            free_func_(entry.cmd_[i]);                \
+        free(entry.cmd_);                             \
+        entry.cmd_ = NULL;                            \
+        entry.cmd_ ## _count = 0;                     \
+    }                                                 \
+    while (0)
+
+    FREE_SEQUENCE(incl, free);
+    FREE_SEQUENCE(reg, cyaml_free_obj);
+    FREE_SEQUENCE(unreg, cyaml_free_obj);
+    FREE_SEQUENCE(add, cyaml_free_inst);
+    FREE_SEQUENCE(get, cyaml_free_inst);
+    FREE_SEQUENCE(delete, cyaml_free_inst);
+    FREE_SEQUENCE(copy, cyaml_free_inst);
+    FREE_SEQUENCE(set, cyaml_free_inst);
+
+#undef FREE_SEQUENCE
+}
+
+void
+cyaml_free_hist_seq(history_seq hist_seq)
+{
+    unsigned int i;
+
+    for (i = 0; i < hist_seq.entries_count; i++)
+        cyaml_free_hist_entry(hist_seq.entries[i]);
+    free(hist_seq.entries);
+    hist_seq.entries = NULL;
```

Comment:
You understand that you're changing the local copy of `hist_seq`, so these changes are not visible outside?

Reply:
Shame on me. Fixed.